File: src/sdk/reporter.js
// agennect-open SDK — Mode B invocation reporter (JavaScript / Node 20+)
//
// Copy this file into your agent or orchestrator project. No dependencies
// beyond global fetch (Node 18+, modern browsers).
//
// Usage:
//   import { AgennectReporter } from './reporter.js';
//
//   const reporter = new AgennectReporter('http://localhost:3000', {
//     callerId: 'my-orchestrator'
//   });
//
//   // Option A: let the SDK invoke the agent and report for you.
//   const { invoke } = reporter.wrap('dataoracle-x7k2');
//   const { result, latency_ms } = await invoke(
//     'https://myagent.com/tasks',
//     { message: { parts: [{ type: 'text', text: 'Analyze Q1 data' }] } }
//   );
//
//   // Option B: report metrics for an invocation you handled yourself.
//   await reporter.report('dataoracle-x7k2', {
//     latency_ms: 340,
//     status: 'success',
//     request_size: 512,
//     response_size: 1024
//   });

export class AgennectReporter {
  constructor(registryUrl, options = {}) {
    if (!registryUrl) throw new Error('AgennectReporter: registryUrl is required');
    this.registryUrl = String(registryUrl).replace(/\/$/, '');
    this.reportTimeoutMs = options.reportTimeoutMs ?? 5000;
    this.invokeTimeoutMs = options.invokeTimeoutMs ?? 30000;
    this.callerId = options.callerId || null;
  }

  // Returns an object with .invoke() bound to one agent id.
  wrap(agentId) {
    if (!agentId) throw new Error('wrap(): agentId is required');
    return {
      invoke: async (endpointUrl, payload, authHeaders = {}) => {
        return this.invoke(agentId, endpointUrl, payload, authHeaders);
      }
    };
  }

  // Calls the agent endpoint, reports metrics (fire-and-forget),
  // and returns { result, latency_ms }. Throws on non-success.
  async invoke(agentId, endpointUrl, payload, authHeaders = {}) {
    if (!agentId)      throw new Error('invoke(): agentId is required');
    if (!endpointUrl)  throw new Error('invoke(): endpointUrl is required');

    const start = Date.now();
    let status = 'error';
    let errorMsg = null;
    let result = null;
    let responseSize = null;

    const requestBody = JSON.stringify(payload ?? {});
    const requestSize = new TextEncoder().encode(requestBody).length;

    try {
      const res = await fetch(endpointUrl, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json', ...authHeaders },
        body: requestBody,
        signal: AbortSignal.timeout(this.invokeTimeoutMs)
      });

      const text = await res.text();
      responseSize = new TextEncoder().encode(text).length;

      if (res.ok) {
        status = 'success';
        try { result = JSON.parse(text); }
        catch (e) { result = { text }; }
      } else {
        status = 'error';
        errorMsg = `HTTP ${res.status}`;
      }
    } catch (e) {
      status = (e.name === 'TimeoutError' || e.name === 'AbortError')
        ? 'timeout' : 'error';
      errorMsg = e.message;
    }

    const latency_ms = Date.now() - start;

    // Fire-and-forget; don't block the caller on registry availability.
    this._report(agentId, {
      latency_ms,
      status,
      request_size: requestSize,
      response_size: responseSize,
      error_msg: errorMsg
    }).catch((err) => {
      console.error('[AgennectReporter] report failed:', err.message);
    });

    if (status !== 'success') {
      const err = new Error(errorMsg || 'Invocation failed');
      err.status = status;
      err.latency_ms = latency_ms;
      throw err;
    }

    return { result, latency_ms };
  }

  // Report metrics for an invocation you handled yourself.
  async report(agentId, metrics) {
    if (!agentId) throw new Error('report(): agentId is required');
    return this._report(agentId, metrics);
  }

  async _report(agentId, metrics) {
    const headers = { 'Content-Type': 'application/json' };
    if (this.callerId) headers['X-Caller-ID'] = this.callerId;

    const res = await fetch(
      `${this.registryUrl}/agents/${encodeURIComponent(agentId)}/report`,
      {
        method: 'POST',
        headers,
        body: JSON.stringify(metrics),
        signal: AbortSignal.timeout(this.reportTimeoutMs)
      }
    );

    if (!res.ok) {
      const text = await res.text().catch(() => '');
      throw new Error(`Report failed: ${res.status} ${text}`);
    }
    return res.json();
  }
}

File: src/sdk/reporter.py
"""
agennect-open SDK — Mode B invocation reporter (Python).

Copy this file into your agent or orchestrator project. Uses httpx when
available, falls back to urllib.request from the stdlib.

Usage:
    from reporter import AgennectReporter

    reporter = AgennectReporter(
        "http://localhost:3000", caller_id="my-orchestrator"
    )

    # Option A: let the SDK invoke and report for you.
    result = reporter.invoke(
        agent_id="dataoracle-x7k2",
        endpoint_url="https://myagent.com/tasks",
        payload={"message": {"parts": [{"type": "text",
                                         "text": "Analyze Q1 data"}]}}
    )

    # Option B: report manually if you handled the invocation yourself.
    reporter.report(
        agent_id="dataoracle-x7k2",
        latency_ms=340,
        status="success",
        request_size=512,
        response_size=1024,
    )
"""

from __future__ import annotations

import json
import threading
import time
from typing import Any, Dict, Optional
from urllib.parse import quote

try:
    import httpx
    _HTTP = "httpx"
except ImportError:
    import urllib.request
    _HTTP = "urllib"


class AgennectReporter:
    def __init__(
        self,
        registry_url: str,
        caller_id: Optional[str] = None,
        report_timeout: float = 5.0,
        invoke_timeout: float = 30.0,
    ) -> None:
        if not registry_url:
            raise ValueError("registry_url is required")
        self.registry_url = registry_url.rstrip("/")
        self.caller_id = caller_id
        self.report_timeout = report_timeout
        self.invoke_timeout = invoke_timeout

    def invoke(
        self,
        agent_id: str,
        endpoint_url: str,
        payload: Dict[str, Any],
        auth_headers: Optional[Dict[str, str]] = None,
    ) -> Dict[str, Any]:
        """Invoke an agent directly and report metrics. Returns the agent's parsed JSON response."""
        if not agent_id:
            raise ValueError("agent_id is required")
        if not endpoint_url:
            raise ValueError("endpoint_url is required")

        headers = {"Content-Type": "application/json"}
        if auth_headers:
            headers.update(auth_headers)

        body = json.dumps(payload or {}).encode()
        request_size = len(body)

        start = time.time()
        status = "error"
        error_msg: Optional[str] = None
        result: Optional[Dict[str, Any]] = None
        response_size: Optional[int] = None

        try:
            if _HTTP == "httpx":
                with httpx.Client() as client:
                    res = client.post(
                        endpoint_url,
                        content=body,
                        headers=headers,
                        timeout=self.invoke_timeout,
                    )
                    text = res.text
                    response_size = len(text.encode())
                    if res.is_success:
                        status = "success"
                        try:
                            result = res.json()
                        except Exception:
                            result = {"text": text}
                    else:
                        error_msg = f"HTTP {res.status_code}"
            else:
                req = urllib.request.Request(
                    endpoint_url, data=body, headers=headers, method="POST"
                )
                with urllib.request.urlopen(req, timeout=self.invoke_timeout) as res:
                    text = res.read().decode()
                    response_size = len(text.encode())
                    status = "success"
                    try:
                        result = json.loads(text)
                    except Exception:
                        result = {"text": text}
        except Exception as e:
            status = "timeout" if "timeout" in str(e).lower() else "error"
            error_msg = str(e)

        latency_ms = int((time.time() - start) * 1000)

        # Fire-and-forget report so the caller never blocks on registry availability.
        threading.Thread(
            target=self._report_silently,
            args=(agent_id, latency_ms, status, request_size, response_size, error_msg),
            daemon=True,
        ).start()

        if status != "success":
            raise RuntimeError(error_msg or "Invocation failed")

        return result or {}

    def report(
        self,
        agent_id: str,
        latency_ms: int,
        status: str,
        request_size: Optional[int] = None,
        response_size: Optional[int] = None,
        error_msg: Optional[str] = None,
        caller_agent_id: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Manually report an invocation you handled yourself."""
        return self._report(
            agent_id, latency_ms, status,
            request_size, response_size, error_msg, caller_agent_id,
        )

    # ── internal ────────────────────────────────────────────────────────────

    def _report(
        self,
        agent_id: str,
        latency_ms: int,
        status: str,
        request_size: Optional[int],
        response_size: Optional[int],
        error_msg: Optional[str],
        caller_agent_id: Optional[str] = None,
    ) -> Dict[str, Any]:
        url = f"{self.registry_url}/agents/{quote(agent_id, safe='')}/report"
        headers = {"Content-Type": "application/json"}
        if self.caller_id:
            headers["X-Caller-ID"] = self.caller_id

        body = json.dumps({
            "latency_ms":      latency_ms,
            "status":          status,
            "request_size":    request_size,
            "response_size":   response_size,
            "error_msg":       error_msg,
            "caller_agent_id": caller_agent_id,
        }).encode()

        if _HTTP == "httpx":
            with httpx.Client() as client:
                res = client.post(
                    url, content=body, headers=headers,
                    timeout=self.report_timeout,
                )
                res.raise_for_status()
                return res.json()
        else:
            req = urllib.request.Request(url, data=body, headers=headers, method="POST")
            with urllib.request.urlopen(req, timeout=self.report_timeout) as res:
                return json.loads(res.read().decode())

    def _report_silently(self, *args) -> None:
        try:
            self._report(*args)
        except Exception as e:
            # Reporting must never crash the caller.
            print(f"[AgennectReporter] report failed: {e}")
